### PR TITLE
remove intrm_tables，drop change to delete by walk.

### PR DESCRIPTION
**Pull request by ZR233**
When operating page tables with memory allocators, Vec that relies on global allocators should not be used, otherwise deadlocks may occur.

### CODE REVIEW

File: page_table_multiarch/src/bits64.rs [hunk 1]
```diff
@@ -379,8 +376,15 @@ impl<M: PagingMetaData, PTE: GenericPTE, H: PagingHandler> PageTable64<M, PTE, H
 
 impl<M: PagingMetaData, PTE: GenericPTE, H: PagingHandler> Drop for PageTable64<M, PTE, H> {
     fn drop(&mut self) {
-        for frame in &self.intrm_tables {
-            H::dealloc_frame(*frame);
+        // don't free the entries in last level, they are not array.
```
Comment (equation314) [inline page_table_multiarch/src/bits64.rs:379]: Why not walk page table only once and check if `level < M::LEVEL - 1`?

File: page_table_multiarch/Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "page_table_multiarch"
-version = "0.1.2"
+version = "0.1.3"
```
Comment (equation314) [inline page_table_multiarch/Cargo.toml:3]: I mean the version number increment should not be included in this PR, there may be other changes before the next release.

File: page_table_multiarch/src/bits64.rs (hunk 1 above)
Comment (ZR233) [reply page_table_multiarch/src/bits64.rs:379]: Because the walk starts from level0, deleting at this level would prevent traversal of the next lower level. Therefore, deletion needs to proceed from higher levels down to lower levels.

Comment (equation314) [reply page_table_multiarch/src/bits64.rs:379]: You can add another `func` arguments in `walk` and `walk_recursive` to use post-order traversal.
Walk the page table three times are too expensive.

File: page_table_multiarch/src/bits64.rs [hunk 3]
```diff
@@ -379,8 +381,13 @@ impl<M: PagingMetaData, PTE: GenericPTE, H: PagingHandler> PageTable64<M, PTE, H
 
 impl<M: PagingMetaData, PTE: GenericPTE, H: PagingHandler> Drop for PageTable64<M, PTE, H> {
     fn drop(&mut self) {
-        for frame in &self.intrm_tables {
-            H::dealloc_frame(*frame);
-        }
+        // don't free the entries in last level, they are not array.
+        let _ = self.walk(usize::MAX, true, &|level, _index, _vaddr, entry| {
+            if level == M::LEVELS - 1 || !entry.is_present() || entry.is_huge() {
```
Comment (equation314) [inline page_table_multiarch/src/bits64.rs:386]: Use `if xx && xx && xx { H::dealloc_frame(entry.paddr()) }` can be more cleaner.

**Review by equation314 — APPROVED** [review]
